pkg/mcs/tso/server/server.go

```diff
@@ -250,8 +430,155 @@ func checkStream(streamCtx context.Context, cancel context.CancelFunc, done chan
 }
 
 // GetTLSConfig get the security config.
-// TODO: implement it
 func (s *Server) GetTLSConfig() *grpcutil.TLSConfig {
+	return &s.cfg.Security.TLSConfig
+}
+
+func (s *Server) initClient() error {
+	tlsConfig, err := s.cfg.Security.ToTLSConfig()
+	if err != nil {
+		return err
+	}
+	u, err := types.NewURLs(strings.Split(s.cfg.BackendEndpoints, ","))
+	if err != nil {
+		return err
+	}
+	s.backendUrls = []url.URL(u)
+	s.etcdClient, s.httpClient, err = etcdutil.CreateClients(tlsConfig, s.backendUrls)
+	return err
+}
+
+func (s *Server) startGRPCServer(l net.Listener) {
+	defer s.serverLoopWg.Done()
+
+	gs := grpc.NewServer()
+	s.service.RegisterGRPCService(gs)
+	serverr := gs.Serve(l)
+
+	log.Info("grpc server stopped serving.", errs.ZapError(serverr))
+	log.Info("try to gracefully stop the server now.")
+
+	// Attempt graceful stop (waits for pending RPCs), but force a stop if
+	// it doesn't happen in a reasonable amount of time.
+	done := make(chan struct{})
+	go func() {
+		gs.GracefulStop()
+		close(done)
+	}()
+	select {
+	case <-done:
+	case <-time.After(defaultGRPCGracefulStopTimeout):
+		log.Info("stopping grpc gracefully is taking longer than expected and force stopping now.")
+		gs.Stop()
+	}
+
+	if s.IsClosed() {
+		log.Info("grpc server stopped.")
+	} else {
+		log.Fatal("grpc server stopped unexpectedly.", errs.ZapError(serverr))
+	}
+}
+
+func (s *Server) startHTTPServer(l net.Listener) {
+	defer s.serverLoopWg.Done()
+
+	handler, _ := SetUpRestHandler(s.service)
+	hs := &http.Server{
+		Handler:           handler,
+		ReadTimeout:       5 * time.Minute,
+		ReadHeaderTimeout: 5 * time.Second,
+	}
+	serverr := hs.Serve(l)
+
+	log.Info("http server stopped serving.", errs.ZapError(serverr))
+	log.Info("try to gracefully shutdown the server now.")
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultHTTPGracefulShutdownTimeout)
+	defer cancel()
+	err := hs.Shutdown(ctx)
+	log.Info("all http(s) requests finished.")
+	if err != nil {
+		log.Error("http server shutdown encountered problem.", errs.ZapError(err))
+	}
+
+	if s.IsClosed() {
+		log.Info("http server stopped.")
+	} else {
+		log.Fatal("http server stopped unexpectedly.", errs.ZapError(serverr))
```

Comment:
ditto

Reply:
as above. merged the above two log lines into one "log.Info("http server stopped serving. try to gracefully shutdown the server now", errs.ZapError(serverr))"

Reply:
I think we don't need to output errors when expertly closing server.

Reply:
how about outputting errors for now? I want to check in the production env, when the exit signal is fired, what are the possible error codes returned from grpcServer.Serv, httpServer.Serv and mux.Serv. We can remove the error code when it's a normal exit after fully understanding the behavior.

Reply:
removed the error code when exit normally